Add link caching using SQLite

saswatamcode opened the pull request:
This PR adds link caching using a local SQLite file, `.mdoxcache` which is suitable for GitHub Actions. It introduces the following:

Adds following fields to `links.validate.config`:
- `cacheType`:  Specifies type of cache. Can be set to 'SQLite' or 'None'. 'SQLite' creates a database in `.mdoxcache`. 'None' or setting it to empty will ensure no `.mdoxcache` file is created. Will throw error if set to anything else.
- `cacheValidity` which specifies duration till which link in the cache database is valid. Default is `120h` or 5 days.

Also adds following flag in case cache needs to be cleared but there is no filesystem access:
- `--cache.clear` which instructs mdox to drop the entire cache database and start afresh.

bwplotka commented on pkg/cache/cache.go:
```diff
@@ -0,0 +1,160 @@
+// Visited inserts new URL to cache database.
+func (s *Storage) Visited(requestID uint64) error {
+	// If particular URL is already inserted, then delete.
+	// Visit method will only be called if validity expires for a URL or in case of a new URL.
```
I would not assume the usage. Focus on Storage API. What would you expect from `Visited`. I would only expect... No idea, `visited` is the wrong name even as it suggests a bool return if something was visited. If that means just to `visit` something then we should probably bump the timestamp, nothing else, no?

saswatamcode replied:
Yes, `Visited` seems like the wrong name for this but I checked this [here](https://github.com/gocolly/colly/blob/master/colly.go#L755) and [here](https://github.com/velebak/colly-sqlite3-storage/blob/master/colly/sqlite3/sqlite3.go#L111), so seems like it actually inserts it into storage. So had to do it this way.

bwplotka commented on pkg/cache/cache.go:
```diff
@@ -0,0 +1,160 @@
+	}
+
+	// Check if URL is within validity threshold.
+	now := time.Now().UTC()
+	if !timestamp.IsZero() && now.Sub(timestamp) <= s.Validity {
+		return true, nil
+	}
+	return false, nil
+}
```
```suggestion
	return (!timestamp.IsZero() && ime.Now().UTC().Sub(timestamp) <= s.Validity), nil
}
```

bwplotka commented on pkg/mdformatter/linktransformer/link.go:
```diff
@@ -180,6 +190,20 @@ func NewValidator(ctx context.Context, logger log.Logger, linksValidateConfig []
 		v.rMu.Lock()
 		defer v.rMu.Unlock()
+		// Colly stores requests in cache, even when they have errors.
+		// So need to delete requests with errors from cache database.
```
OMG this is quite annoying, can we add an issue on Colly to propose a different API?

saswatamcode replied:
Yes, I plan to add a few issues. 🙂

bwplotka commented on pkg/mdformatter/linktransformer/link.go:
```diff
@@ -180,6 +190,20 @@ func NewValidator(ctx context.Context, logger log.Logger, linksValidateConfig []
+		// Colly stores requests in cache, even when they have errors.
+		// So need to delete requests with errors from cache database.
+		if v.storage != nil {
+			h := fnv.New64a()
```
Let's add that to separate method, ideally on storage side? (hashing request).

Are you sure it's the same as colly hash. What if they change that next version (quite prone to errro code)

saswatamcode replied:
I wanted to do it on storage side too, but have to conform colly storage interface so that we can pass that in.

bwplotka commented on pkg/mdformatter/linktransformer/link.go:
```diff
@@ -180,6 +190,20 @@ func NewValidator(ctx context.Context, logger log.Logger, linksValidateConfig []
+			if writeErr != nil {
+				v.remoteLinks[response.Ctx.Get(originalURLKey)] = errors.Wrapf(writeErr, "%q could not build requestID %v", response.Request.URL.String(), response.StatusCode)
+			}
+			delErr := v.storage.DeleteRequest(h.Sum64())
```
I wonder if it makes sense to use their interface if we have to do all those things manually.. Wouldn't be better to enhance our storage with API like error and store the status code etc too? Knowing fast that something is 404 is also coachable, no? (with maybe different validity)

saswatamcode replied:
Yup. Was kind of thinking about the same thing, maybe would be better to just do this on our own and not pass any special storage to Colly. 🙂

bwplotka reviewed (commented):
Nice, great job! Commented some ideas.

In terms of testing, I would test things with and without such cache and use SQL lite file directly, any blocker for that?

saswatamcode commented on README.md:
```diff
@@ -69,6 +69,11 @@ Flags:
                                  flag (mutually exclusive). Content of YAML file
                                  for skipping link check, with spec defined in
                                  github.com/bwplotka/mdox/pkg/linktransformer.ValidatorConfig
+      --no-cache                 If true, link checking will not create local
```
Sure this can be done! 🙂

saswatamcode commented on main.go:
```diff
@@ -125,10 +128,14 @@ This directive runs executable with arguments and put its stderr and stdout outp
+	cacheDisabled := cmd.Flag("no-cache", "If true, link checking will not create local SQLite cache file").Bool()
+	// Default validity is for 5 days.
+	cacheValidity := cmd.Flag("cache.validity", "Duration till which link in cache is considered valid").Default("120h").Duration()
+	clearCache := cmd.Flag("cache.clear", "If true, entire cache database will be dropped").Bool()
```
Yes locally, but I guess we cannot clear GitHub Actions cache yet since this is open: https://github.com/actions/cache/issues/2. So flag might be useful if we want to delete cache immediately from GH action and not wait till GitHub refreshes on its own. 🙂

saswatamcode replied:
Sure!

saswatamcode commented on pkg/cache/cache.go:
```diff
@@ -0,0 +1,160 @@
+
+// Close cache database.
+func (s *Storage) Close() error {
+	err := s.dbHandle.Close()
```
I guess we can remove!

saswatamcode commented on pkg/cache/cache.go:
```diff
@@ -0,0 +1,160 @@
+}
+
+// Visited inserts new URL to cache database.
+func (s *Storage) Visited(requestID uint64) error {
```
This is how colly does it (by hashing URLs for get requests). Seems weird to me too but cannot change as we need to conform to [Storage interface](https://github.com/gocolly/colly/blob/master/storage/storage.go#L30). They hash it [here](https://github.com/gocolly/colly/blob/master/colly.go#L735). 🙂

bwplotka commented on pkg/cache/cache.go:
```diff
@@ -0,0 +1,160 @@
+	if err != nil {
+		return err
+	}
+	statement, err = s.dbHandle.Prepare("CREATE INDEX IF NOT EXISTS idx_visited ON visited (requestID)")
```
Done 
![image](https://user-images.githubusercontent.com/6950331/123811306-0b187b00-d8eb-11eb-90f4-ff0d2278fd7e.png)

saswatamcode replied:
Thank you!! ❤️

bwplotka commented:
```
#TODO(bwplotka): Uncomment once we upstream merrors package: @/tmp/.bin/faillint-v1.5.0 -paths "errors=github.com/pkg/errors" ./...
Error: /home/runner/work/mdox/mdox/pkg/mdformatter/linktransformer/link_test.go:225:50: too few arguments in call to MustNewValidator
faillint: error during loading
```
🤔 Hmmm, let me check today/Monday

bwplotka commented:
Let's try rebasing, maybe there is some old dep? (:

saswatamcode commented:
Yup! Rebasing fixed this!

bwplotka commented on pkg/cache/cache.go:
```diff
@@ -0,0 +1,143 @@
+// Copyright (c) Bartłomiej Płotka @bwplotka
+// Licensed under the Apache License 2.0.
+
+package cache
```
It would be amazing to get some feedback to their project (just mention, don't expect anything).

bwplotka commented on .mdox.validate.yaml:
```diff
@@ -1,4 +1,5 @@
 version: 1
+noCache: true
```
Again, let's avoid boolean flags. What if we have another cache soon with different type? I believe there should be `cacheType` or something (: e.g `SQlite` and `None`

saswatamcode replied:
Sure! Will implement this. 🙂

saswatamcode commented on .gitignore:
```diff
@@ -16,4 +16,5 @@ vendor/
-.envrc
\ No newline at end of file
+.envrc
+mdoxcache
```
Sure!

matej-g commented on pkg/mdformatter/linktransformer/link.go:
```diff
@@ -197,9 +204,13 @@ type futureResult struct {
+func NewValidator(ctx context.Context, logger log.Logger, linksValidateConfig []byte, anchorDir string, storage *cache.Storage, reg *prometheus.Registry) (mdformatter.LinkTransformer, error) {
 	var err error
-	config := Config{}
+	defaultValidity, err := time.ParseDuration("120h")
```
Why not directly create a `const` with default `time.Duration`? It would save us from extra parsing here :blush:

saswatamcode replied:
Will do that!

bwplotka replied:
still not done?

saswatamcode replied:
Yes, needs to be done!

matej-g commented on pkg/mdformatter/linktransformer/link.go:
```diff
@@ -239,6 +251,15 @@ func NewValidator(ctx context.Context, logger log.Logger, linksValidateConfig []
 		v.c.SetRequestTimeout(config.timeout)
 	}
+
+	if v.validateConfig.CacheType == sqlite && storage != nil {
```
Do we need the cache type check? :thinking: As long as `storage` is not `nil`, it should not matter what type we're initializing, right?

matej-g replied:
Hm then I would maybe expect `storage` to still be `nil`, after all if we do not want to initialize storage why would it not be `nil`? I think it's detail, it just feels like it could be simplified a bit, but it's more of a nit.

bwplotka replied:
still to be done?

saswatamcode replied:
We can iterate on this in separate PR!

matej-g commented on pkg/cache/cache.go:
```diff
@@ -0,0 +1,143 @@
+// Init initializes cache database.
+func (s *Storage) Init() error {
+	// Check if db exists.
+	if s.dbHandle == nil {
```
I'm wondering, if the database exists, do we need to go through the other steps below? Could we simply check if the database and table exists (optionally, truncate the table if `s.ClearCache` true) and if yes, exist, otherwise create new table? Would that flow make sense?

saswatamcode replied:
Let me try that out and get back! 🙂

matej-g reviewed (commented):
Wonderful job as usual :hugs: I left some comments, would be cool to get it merged soon!

matej-g commented:
I'm wondering what's status of this PR? It was looking good enough for merging for me but I'm wondering if @bwplotka had a chance to look at the latest state as well? cc @saswatamcode

saswatamcode commented:
I didn't get a chance to look through the `dbHandle` flow that was mentioned in the last review. 😅
I think this is in a mergeable state once @bwplotka has had a look and we can iterate on this (will possibly need to once we use it in a repo and come across some new issues anyway), @matej-g ! 🙂

bwplotka commented on pkg/mdformatter/linktransformer/link.go:
```diff
@@ -239,6 +251,15 @@ func NewValidator(ctx context.Context, logger log.Logger, linksValidateConfig []
+
+	if v.validateConfig.CacheType == sqlite && storage != nil {
+		v.storage = storage
+		v.storage.Validity = v.validateConfig.CacheValidity
```
This looks like a code smell.

We pass storage here for some composability, but we kind of  use it like struct (setting Field directly) despite doing  `Init`  below. Can we add `Validity` parameter to `Init` then?

saswatamcode replied:
Will try to restructure this!

bwplotka commented on pkg/mdformatter/linktransformer/link.go:
```diff
@@ -253,6 +274,12 @@ func NewValidator(ctx context.Context, logger log.Logger, linksValidateConfig []
 		v.rMu.Lock()
 		defer v.rMu.Unlock()
+		if v.storage != nil {
+			err := v.storage.CacheURL(response.Ctx.Get(originalURLKey))
+			if err != nil {
```
```suggestion
			if err := v.storage.CacheURL(response.Ctx.Get(originalURLKey)); err != nil {
```

bwplotka commented on pkg/cache/cache.go:
```diff
@@ -0,0 +1,143 @@
+		return false, err
+	}
+
+	// Check if URL is within validity threshold.
```
I would add some jitter to it (and jitter might be parameter). The reason is that we will have timeouts still if we will expire all items in one go.

saswatamcode replied:
Will do!

bwplotka reviewed (approved):
Nice! IMO it's good to be merged, just minor things (: 

Thanks!

saswatamcode commented on main.go:
```diff
@@ -207,10 +210,11 @@ This directive runs executable with arguments and put its stderr and stdout outp
 	linksValidateEnabled := cmd.Flag("links.validate", "If true, all links will be validated").Short('l').Bool()
 	linksValidateConfig := extflag.RegisterPathOrContent(cmd, "links.validate.config", "YAML file for skipping link check, with spec defined in github.com/bwplotka/mdox/pkg/linktransformer.ValidatorConfig", extflag.WithEnvSubstitution())
 
+	clearCache := cmd.Flag("cache.clear", "If true, entire cache database will be dropped. Useful in case mdox cache needs to be cleared immediately from GitHub Actions or other CI runner cache.").Bool()
```
Will update!

bwplotka commented:
Still to be addressed I believe (:

matej-g commented:
Any chance of progress? Before this PR celebrates it's 1st birthday in a week :birthday: :joy:

saswatamcode commented:
Oh no! Yup, let me get to it soon, it's been on my backlog for a long time now! 🙈

saswatamcode commented:
@bwplotka @matej-g I've addressed most of the feedback here. But yes, there are some code smells due to cache, that I think arise from not having a dedicated `Cache` interface. We try to support multiple `CacheType`s here but use only the sqlite `Storage` struct, which makes it a bit confusing. 
Let's get this merged, I will refactor this in another PR, without changing configuration semantics. 🙂

bwplotka commented on pkg/cache/cache.go:
```diff
@@ -0,0 +1,150 @@
+)
+
+// Storage implements a SQLite3 caching backend for Colly.
+type Storage struct {
```
Sqlite3Storage

bwplotka commented on pkg/cache/cache.go:
```diff
@@ -0,0 +1,150 @@
+		return false, err
+	}
+
+	// Check if URL is within validity threshold with jitter.
```
If we want jitter, let's randomize it from [0 -> Validity] (:

bwplotka reviewed (approved):
Amazing, thanks! Small nits only.

douglascamata commented on pkg/mdformatter/linktransformer/config.go:
```diff
@@ -69,14 +77,19 @@ const (
+	none   = "None"
+	sqlite = "SQLite"
+
+	defaultValidity = time.Duration(432000000000000)
```
Better to convert this duration into something more readable: 
```suggestion
	defaultValidity = 2 * time.Hour
```

douglascamata replied:
Could even create a separate constant for the duration of a day (there's no `time.Day`) to make it better.

douglascamata commented on pkg/cache/cache.go:
```diff
@@ -0,0 +1,158 @@
+	if s.Jitter == time.Duration(0) {
+
+		// In case a custom jitter is not provided, a jitter from 0 to cache validity is used.
+		return (!timestamp.IsZero() && time.Now().UTC().Sub(timestamp)+time.Duration(rand.Intn(int(s.Validity))) <= s.Validity), nil
```
Another tip, `time.Now().UTC().Sub(timestamp)` is the same as `time.Since(timestamp)`.

bwplotka commented on pkg/cache/cache.go:
```diff
@@ -0,0 +1,158 @@
+	}
+
+	// Check if URL is within validity threshold with jitter.
+	rand.Seed(time.Now().UnixNano())
```
This should be done in the init of struct and ideally on non-global rand if possible (:

bwplotka commented on pkg/cache/cache.go:
```diff
@@ -0,0 +1,158 @@
+	if s.Jitter == time.Duration(0) {
+
+		// In case a custom jitter is not provided, a jitter from 0 to cache validity is used.
+		return (!timestamp.IsZero() && time.Now().UTC().Sub(timestamp)+time.Duration(rand.Intn(int(s.Validity))) <= s.Validity), nil
```
Plus I think it might make sense to just have one comparison with `jitterValue` being 0 for jitter equal 0 and value rand[0-jitter] otherwise (:

bwplotka reviewed (approved):
Amazing.

+1 to @douglascamata comments, plus added few more. Thanks!

bwplotka reviewed (approved):
Tiny nit, not a blocker, LGTM! Thanks
